Manage a Pledge: Enable adding/removing contributors from wp-admin.

ryelle opened the pull request:
This updates the display of contributors into a table view, and adds the ability to add and remove contributors to existing pledges.

The display has been refactored to use JS templates & JSON contributor data– the data is output onto the page when loaded from the server, and rendered when the page finishes loading. Adding & removing contributors now submits to an admin-ajax.php endpoint, which, if successful, return the new list of contributors. This ensures the display is always up to date. 

Fixes #3.

<img width="483" alt="Screen Shot 2019-11-13 at 4 54 19 PM" src="https://user-images.githubusercontent.com/541093/68807714-50186c80-0636-11ea-8bfe-380b5ae82c45.png">

**To test**

- Edit a pledge in wp-admin
- View the contributors, they should be in the correct "confirmed"/"unconfirmed" table
- Resend a confirmation email, it should be sent & you'll get a success alert.
- Remove a contributor, you will have to confirm that you intend to do this.
- Hit cancel, it should not remove the contributor.
- Remove a contributor, hit OK this time
- The list should refresh and that contributor should be gone now
- Add some users to the new contributors text input
- Submit either by hitting enter in the input, or clicking the buttton
- It should succeed, and show those contribs in the unconfirmed list now
- Invalid users should trigger an error message & not add anyone

cc @melchoyce to check design :)

ryelle committed:
Style the email status indicator

ryelle committed:
Style contributors list (first pass)

ryelle committed:
Use JS + AJAX to resend the confirmation email

ryelle committed:
Hook up "Remove" button

ryelle committed:
Add the "Add Contributors" input to the form

ryelle committed:
Refactor to use JSON + templates for contributors

This will make it easier to show contributors as they're added

ryelle committed:
Enable "add contributors" functionality

ryelle committed:
Update display to use new design

ryelle committed:
Fix unescaped content

ryelle committed:
Tweak design

ryelle committed:
Add `wp-util` as a dependency

ryelle committed:
Support the unlikely case of no contributors

ryelle committed:
Indicate to screen readers that the message section will update

coreymckrill reviewed (commented):
Everything worked as expected in testing. 👍 

Two questions:

1. I noticed that when adding new contributors in wp-admin, a confirmation email is not sent once they are successfully added. I'm wondering if it should be sent. My expectation as the user managing the pledge was that it would get sent (the "resend" button is misleading in this case). There would probably need to be some logic there so it only gets sent if the pledge is already published.
1. Security-wise, the handler is checking the ajax referrer, but I wonder if we should also do some kind of permissions check. On the wp-admin side, that would probably be a `current_user_can` check. But if we're going to use the same Ajax for the front end manage form, where there isn't necessarily a logged in user, it would need to somehow verify the access key or something.

ryelle committed:
phpcs fixes

issues introduced in this PR

ryelle committed:
Send contributors the confirmation email when created by the ajax endpoint

This will only send the email for published pledges.

ryelle committed:
Hide the "resend confirmation" button on draft pledges

The confirmation hasn't been sent in the first place, so this button is misleading

ryelle commented:
> a confirmation email is not sent once they are successfully added

Ah, I thought I tested this but I guess I saw the "resent" emails instead. In any case, this has been updated so it only sends emails for contributors to published pledges. I also hid the "resend" button on draft pledges, since they haven't been sent the initial email.

>  On the wp-admin side, that would probably be a current_user_can check. But if we're going to use the same Ajax for the front end manage form, where there isn't necessarily a logged in user, it would need to somehow verify the access key or something.

Good point– it looks like #96 has some auth code that would be useful here, so I'm going to jump over there and pull out the auth code into another PR - this one can either wait for that, or we can merge it first, whichever.

ryelle committed:
Use new auth/email refactored code

ryelle committed:
Endpoints: Update handler to "manage contributors"

ryelle committed:
Import correct namespace, use correct variable

ryelle committed:
Trim stray whitespace
